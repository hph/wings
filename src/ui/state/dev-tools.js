import React from 'react';
import { createDevTools } from 'redux-devtools';
import LogMonitor from 'redux-devtools-log-monitor';
import DockMonitor from 'redux-devtools-dock-monitor';
import Dispatcher from 'redux-devtools-dispatch';
import MultipleMonitors from 'redux-devtools-multiple-monitors';

export default createDevTools(
  <DockMonitor
    toggleVisibilityKey="ctrl-h"
    changePositionKey="ctrl-q"
    changeMonitorKey="ctrl-m"
    defaultIsVisible={false}
  >
    <MultipleMonitors>
      <LogMonitor theme="solarized" />
      <Dispatcher theme="solarized" />
    </MultipleMonitors>
  </DockMonitor> // eslint-disable-line comma-dangle
);
