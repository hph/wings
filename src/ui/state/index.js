export * as actions from 'ui/state/actions';
export * as selectors from 'ui/state/selectors';
export * as types from 'ui/state/types';
export configureStore from 'ui/state/store/shared';
export middleware from 'ui/state/middleware';
export reducer from 'ui/state/reducer';
