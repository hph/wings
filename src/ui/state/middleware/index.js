import { applyMiddleware } from 'redux';

import userInput from 'ui/state/middleware/user-input';
import cursorPosition from 'ui/state/middleware/cursor-position';
import commandBar from 'ui/state/middleware/command-bar';

export default applyMiddleware(
  userInput,
  cursorPosition,
  commandBar,
);
