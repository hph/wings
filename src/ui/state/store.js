import { applyMiddleware, createStore, compose } from 'redux';

import {
  DevTools,
  reducer,
  createUserInputMiddleware,
} from 'ui/state';


export default function configureStore (initialState) {
  let enhancer = applyMiddleware(createUserInputMiddleware());

  if (process.env.NODE_ENV === 'development') {
    enhancer = compose(enhancer, DevTools.instrument({ maxAge: 30 }));
  }

  return createStore(reducer, initialState, enhancer);
}
