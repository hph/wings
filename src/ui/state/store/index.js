const file = process.env.NODE_ENV === 'production'
  ? 'production'
  : 'development'; // Also use this store in the test environment.
// eslint-disable-next-line import/no-dynamic-require
module.exports = require(`./${ file }`).default;
