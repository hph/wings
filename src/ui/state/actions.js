import { types } from 'ui/state';

let nextPaneId = 0;
export function createPane(filename, text) {
  nextPaneId += 1;
  return {
    type: types.CREATE_PANE,
    id: nextPaneId,
    filename,
    text,
  };
}

export function updatePane(id, options) {
  return {
    type: types.UPDATE_PANE,
    id,
    ...options,
  };
}

export function destroyPane(id) {
  return {
    type: types.DESTROY_PANE,
    id,
  };
}

export function updateConfig(options) {
  return {
    type: types.UPDATE_CONFIG,
    ...options,
  };
}

export function updateCommand(options) {
  return {
    type: types.UPDATE_COMMAND,
    ...options,
  };
}

export function execCommand(options) {
  return {
    type: types.EXEC_COMMAND,
    ...options,
  };
}

export function toggleTreeView() {
  return {
    type: types.TOGGLE_TREE_VIEW,
  };
}

export function userInputFocus(focus) {
  return {
    type: focus ? types.USER_INPUT_FOCUS : types.USER_INPUT_UNFOCUS,
  };
}
