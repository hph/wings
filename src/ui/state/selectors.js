import _ from 'lodash';
import { createSelector } from 'reselect';

import { computeFontDimensions } from 'ui/utils';

export const currentPane = createSelector(
  state => state.panes,
  state => state.config.currentPaneId,
  (panes, currentPaneId) => currentPaneId
    ? _.find(panes, ({ id }) => id === currentPaneId)
    : panes[0],
);

export const paneById = createSelector(
  state => state.panes,
  (state, props) => props.paneId,
  (panes, paneId) => _.find(panes, ({ id }) => id === paneId),
);

export const charSizes = createSelector(
  state => state.config.theme,
  theme => {
    const { width, height } = computeFontDimensions(theme);
    return {
      charHeight: height,
      charWidth: width,
    };
  },
);
