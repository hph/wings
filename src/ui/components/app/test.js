import React from 'react';
import renderer from 'react-test-renderer';
import setCustomProperties from 'dynamic-css-properties';

import { updateConfig } from 'ui/state/actions';
import {
  App,
  setTheme,
  isFullscreen,
  mapDispatchToProps,
  mapStateToProps,
} from './index';

jest.mock('dynamic-css-properties', () => jest.fn());
jest.mock('ui/components/browser', () => 'Browser');
jest.mock('ui/components/command-bar', () => 'CommandBar');
jest.mock('ui/components/logo', () => 'Logo');
jest.mock('ui/components/title-bar', () => 'TitleBar');
jest.mock('ui/components/tree-view', () => 'TreeView');
jest.mock('ui/components/pane', () => 'Pane');
jest.mock('ui/utils', () => ({
  computeFontDimensions: jest.fn(() => ({ width: 5, height: 21 })),
}));
jest.mock('ui/state/actions', () => ({
  updateConfig: jest.fn(),
}));

const defaultProps = {
  isBrowserVisible: false,
  isTitleBarVisible: false,
  isTreeViewVisible: false,
  isFullscreen: jest.fn(false),
  showTitleBar: jest.fn(),
  setTheme: jest.fn(),
  isCommandBarVisible: false,
  charWidth: 5,
  theme: {
    titleBarHeight: 23,
  },
  panes: [],
};

describe('App', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const createSnapshot = passedProps => {
    const props = {
      ...defaultProps,
      ...passedProps,
    };
    expect(renderer.create(<App {...props} />).toJSON()).toMatchSnapshot();
  };

  it('defaults to rendering a logo in an empty state', () => {
    createSnapshot();
  });

  it('renders a TitleBar when so configured', () => {
    createSnapshot({
      isTitleBarVisible: true,
    });
  });

  it('renders a CommandBar when so configured', () => {
    createSnapshot({
      isCommandBarVisible: true,
    });
  });

  it('renders a TreeView when so configured', () => {
    createSnapshot({
      isTreeViewVisible: true,
    });
  });

  it('renders a Browser when so configured', () => {
    createSnapshot({
      isBrowserVisible: true,
    });
  });

  it('renders a Pane when so configured', () => {
    createSnapshot({
      panes: [{ id: 1 }],
    });
  });

  it('renders multiple instances of Pane when so configured', () => {
    createSnapshot({
      panes: [{ id: 1 }, { id: 2 }],
    });
  });

  it('can render many things at once when so configured', () => {
    createSnapshot({
      isBrowserVisible: true,
      isTitleBarVisible: true,
      isTreeViewVisible: true,
      isCommandBarVisible: true,
      panes: [{ id: 1 }, { id: 2 }],
    });
  });

  it('will call setTheme with the provided props when instantiated', () => {
    new App(defaultProps); // eslint-disable-line no-new
    expect(defaultProps.setTheme).toHaveBeenCalledWith(defaultProps);
  });

  it('will set a global resize event listener on mount', () => {
    window.addEventListener = jest.fn();
    const app = new App(defaultProps);
    app.componentDidMount();

    expect(window.addEventListener).toHaveBeenCalledWith(
      'resize',
      app.onResize,
    );
  });

  it('calls setTheme in componentWillReceiveProps when the theme changes', () => {
    const app = new App(defaultProps);
    defaultProps.setTheme.mockReset();
    app.componentWillReceiveProps(defaultProps);
    expect(defaultProps.setTheme).not.toHaveBeenCalled();

    const nextProps = {
      ...defaultProps,
      theme: {
        color: 'hotpink',
      },
    };
    app.componentWillReceiveProps(nextProps);
    expect(defaultProps.setTheme).toHaveBeenCalledWith(nextProps);
  });

  it('should call showTitleBar in onResize', () => {
    const app = new App(defaultProps);
    app.showOrHideTitleBar = jest.fn();
    app.onResize();

    expect(app.showOrHideTitleBar).toHaveBeenCalledWith(expect.any(Number));
  });
});

describe('App showOrHideTitleBar', () => {
  it('should exit if more than 150 ms have passed', () => {
    window.requestAnimationFrame = jest.fn();
    const app = new App(defaultProps);

    app.showOrHideTitleBar(new Date().getTime() - 200);
    // This function is always called if it does not exit.
    expect(window.requestAnimationFrame).not.toHaveBeenCalled();

    app.showOrHideTitleBar(new Date().getTime());
    expect(window.requestAnimationFrame).toHaveBeenCalled();
  });

  it('should update the visiblity of the titlebar as required', () => {
    window.requestAnimationFrame = jest.fn();
    const app = new App({
      ...defaultProps,
      isTitleBarVisible: false,
    });

    app.showOrHideTitleBar(new Date().getTime());

    window.innerHeight = 0;
    window.screen.height = 1;
    app.showOrHideTitleBar(new Date().getTime());
    expect(defaultProps.showTitleBar).toHaveBeenCalledWith(true);
  });

  it('should update css variables when the titlebar visibility changes', () => {
    window.requestAnimationFrame = jest.fn();
    let app = new App(defaultProps);
    defaultProps.setTheme.mockReset();

    window.innerHeight = 0;
    window.screen.height = 0;
    app.showOrHideTitleBar(new Date().getTime());
    expect(defaultProps.setTheme).toHaveBeenCalledWith({
      charWidth: 5,
      isTitleBarVisible: true,
      theme: {
        titleBarHeight: 23,
      },
    });

    window.innerHeight = 0;
    window.screen.height = 1;
    defaultProps.setTheme.mockReset();
    app.showOrHideTitleBar(new Date().getTime());
    expect(defaultProps.setTheme).toHaveBeenCalled();

    const props = {
      ...defaultProps,
      isTitleBarVisible: true,
    };
    app = new App(props);

    window.innerHeight = 0;
    window.screen.height = 0;
    app.showOrHideTitleBar(new Date().getTime());
    expect(props.setTheme).toHaveBeenCalledWith(props);
  });

  it('should render an error boundary when componentDidCatch is fired', () => {
    const app = new App(defaultProps);
    app.setState = jest.fn();
    const error = new Error('some error');
    const info = { componentStack: 'some info' };
    app.componentDidCatch(error, info);
    expect(app.setState).toHaveBeenCalledWith({
      error,
      errorStack: info.componentStack,
    });
  });

  it('should render an error boundary when there is an error', () => {
    const component = renderer.create(<App {...defaultProps} />);
    expect(component.toJSON()).toMatchSnapshot();
    component.getInstance().setState({
      error: new Error('an error'),
      errorStack: 'some stack info',
    });
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('setTheme', () => {
  it('should call setCustomProperties with options computed from the provided arguments', () => {
    setTheme({
      charWidth: 5,
      isTitleBarVisible: true,
      theme: {
        titleBarHeight: 23,
      },
    });
    expect(setCustomProperties).toHaveBeenCalledWith({
      charWidth: '5px',
      titleBarHeight: 23,
      viewportHeight: 'calc(100vh - var(--title-bar-height)',
    });
  });

  it('should pass on other theme properties', () => {
    setTheme({
      charWidth: 5,
      isTitleBarVisible: false,
      theme: {
        titleBarHeight: 23,
        anotherThemeVariable: 'red',
      },
    });
    expect(setCustomProperties).toHaveBeenCalledWith(
      expect.objectContaining({
        anotherThemeVariable: 'red',
      }),
    );
  });
});

describe('isFullscreen', () => {
  it('should returns the equality of the inner and screen height', () => {
    window.innerHeight = 10;
    window.screen.height = 20;
    expect(isFullscreen()).toBe(false);

    window.innerHeight = 10;
    window.screen.height = 10;
    expect(isFullscreen()).toBe(true);
  });
});

describe('App mapStateToProps', () => {
  it('mapStateToProps returns props based on state', () => {
    expect(
      mapStateToProps({
        config: {
          isBrowserVisible: true,
          isTreeViewVisible: true,
          isTitleBarVisible: true,
          mode: 'normal',
          theme: {},
        },
        panes: [],
      }),
    ).toEqual({
      charWidth: 5,
      isBrowserVisible: true,
      isTreeViewVisible: true,
      isTitleBarVisible: true,
      isCommandBarVisible: false,
      theme: {},
      panes: [],
    });
  });
});

describe('mapDispatchToProps', () => {
  it('should provide various functions as props', () => {
    const funcs = Object.keys(mapDispatchToProps());
    expect(funcs).toEqual(['setTheme', 'isFullscreen', 'showTitleBar']);
  });

  it('should provide a showTitleBar function that wraps updateConfig', () => {
    const dispatch = jest.fn();
    const { showTitleBar } = mapDispatchToProps(dispatch);
    showTitleBar(true);
    expect(updateConfig).toHaveBeenCalledWith({ isTitleBarVisible: true });
    expect(dispatch).toHaveBeenCalled();
  });
});
