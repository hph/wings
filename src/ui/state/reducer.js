import _ from 'lodash';
import { combineReducers } from 'redux';

import fixedKeys from 'ui/fixed-keys';
import * as types from 'ui/state/types';

export function configReducer(state = {}, action) {
  const { type, ...values } = action;
  switch (type) {
    case types.UPDATE_CONFIG: {
      return {
        ...state,
        ...values,
      };
    }

    case types.TOGGLE_TREE_VIEW: {
      return {
        ...state,
        isTreeViewVisible: !state.isTreeViewVisible,
      };
    }

    default:
      return state;
  }
}

export function panesReducer(state = [], action) {
  const { type, ...values } = action;
  switch (type) {
    case types.CREATE_PANE: {
      const lines = (action.text || '').split('\n').slice(0, -1);
      return [
        ...state,
        _.defaults(
          { ..._.omit(values, ['text']) },
          {
            lines: _.isEmpty(lines) ? [''] : lines,
            column: 0,
            row: 0,
            firstVisibleRow: 0,
            firstVisibleColumn: 0,
            width: window.screen.width,
            height: window.screen.height,
          },
        ),
      ];
    }

    case types.UPDATE_PANE: {
      return _.map(state, pane => {
        if (pane.id === action.id) {
          return _.defaults({ ...values }, pane);
        }
        return pane;
      });
    }

    case types.DESTROY_PANE: {
      return _.reject(state, ({ id }) => id === values.id);
    }

    default:
      return state;
  }
}

export function commandReducer(state = '', action) {
  const { type, value } = action;
  switch (type) {
    case types.UPDATE_COMMAND: {
      if (fixedKeys.has(value)) {
        if (value === 'Space') {
          return `${state} `;
        } else if (value === 'Backspace') {
          return state.substring(0, state.length - 1);
        } else if (value === 'Enter') {
          return '';
        }
        return state;
      }
      return state + value;
    }

    default:
      return state;
  }
}

export default combineReducers({
  config: configReducer,
  panes: panesReducer,
  command: commandReducer,
});
