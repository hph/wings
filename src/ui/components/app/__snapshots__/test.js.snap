// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App can render many things at once when so configured 1`] = `
<div
  className="root"
>
  <TitleBar
    label="Wings"
  />
  <div
    className="main"
  >
    <CommandBar />
    <TreeView />
    <Pane
      isFirst={true}
      paneId={1}
    />
    <Pane
      isFirst={false}
      paneId={2}
    />
    <Browser />
  </div>
</div>
`;

exports[`App defaults to rendering a logo in an empty state 1`] = `
<div
  className="root"
>
  <div
    className="main"
  >
    <Logo />
  </div>
</div>
`;

exports[`App renders a Browser when so configured 1`] = `
<div
  className="root"
>
  <div
    className="main"
  >
    <Logo />
    <Browser />
  </div>
</div>
`;

exports[`App renders a CommandBar when so configured 1`] = `
<div
  className="root"
>
  <div
    className="main"
  >
    <CommandBar />
    <Logo />
  </div>
</div>
`;

exports[`App renders a Pane when so configured 1`] = `
<div
  className="root"
>
  <div
    className="main"
  >
    <Pane
      isFirst={true}
      paneId={1}
    />
  </div>
</div>
`;

exports[`App renders a TitleBar when so configured 1`] = `
<div
  className="root"
>
  <TitleBar
    label="Wings"
  />
  <div
    className="main"
  >
    <Logo />
  </div>
</div>
`;

exports[`App renders a TreeView when so configured 1`] = `
<div
  className="root"
>
  <div
    className="main"
  >
    <TreeView />
    <Logo />
  </div>
</div>
`;

exports[`App renders multiple instances of Pane when so configured 1`] = `
<div
  className="root"
>
  <div
    className="main"
  >
    <Pane
      isFirst={true}
      paneId={1}
    />
    <Pane
      isFirst={false}
      paneId={2}
    />
  </div>
</div>
`;

exports[`App showOrHideTitleBar should render an error boundary when there is an error 1`] = `
<div
  className="root"
>
  <div
    className="main"
  >
    <Logo />
  </div>
</div>
`;

exports[`App showOrHideTitleBar should render an error boundary when there is an error 2`] = `
<div
  className="root"
>
  <div
    className="root"
  >
    <div
      className="type"
    >
      Error
    </div>
    <div>
      an error
    </div>
    <div
      className="info"
    >
      some stack info
    </div>
  </div>
</div>
`;
